### DISCUSSION TITLE
Battery assessment tool

### DISCUSSION THREAD

**Discussion by kryltsov** [amberelectric/public-api#155] [Q&A] [answered]
Hi,

This is my first post here. A while ago I requested AmberElectric to assess the viability of installing Tesla using my actual data and received [a well-formatted document](https://drive.google.com/file/d/1i7HTPmb5JKeZwtY4LQl8Yaztn0bJ6JFf/view?usp=drive_link).

My question is if AmberElectric can open-source this project so we can change contribute and use AmberElectric API to allow anyone to plug their own data at any time and most importantly learn how that modelling works.

I personally want to model my data myself but I am completely newbie and have lots of most-likely dumb questions like:

Do you use SmartShift smarts while modelling or do you simply add exported electricity to the battery until fully charged and then apply discharge when my data shows grid import?

If I will try to model myself so I understand how it works precisely:

Should I take into account the price quoted for installation and calculate the price of each full cycle by dividing battery cost by the amount of how many cycles Tesla supports over a lifetime?

What is the minimum battery must hold and what is max in percentage? I assume the battery can not be discharged to 0% or charged to 100%

How much should I take out when charging and discharging - is it a 10% loss each way?

How to calculate a full cycle - if the battery is 50% charged, we add 50% to it and discharge 50% - should it be counted as 1/2 of a full cycle? 

Thanks
Igor

**Comment by madpilot** [comment, marked as answer]
Hey Igor,

> Do you use SmartShift smarts while modelling or do you simply add exported electricity to the battery until fully charged and then apply discharge when my data shows grid import?

Yep, we run Smartshift for each an every 5 minute interval to generate the report. It takes around 24 hours to crunch the numbers for a 12-month period per customer on a fairly beefy machine. So at this stage, we're unlikely to put it behind an API (DOS risk) or open source it at this point. Just from a maintenance and documentation POV, it would require more resources than we have.

That said, if we can make it a couple of orders of magnitude faster and less resource hungry, we might change our mind.

> Should I take into account the price quoted for installation and calculate the price of each full cycle by dividing battery cost by the amount of how many cycles Tesla supports over a lifetime?

I didn't write the tool that generates these reports, but as far as I know, we don't (mainly, because we don't know that data). Smartshift has internal constraints on charge and discharge cycles and tries to limit it to 1 full cycle per day.

> What is the minimum battery must hold and what is max in percentage? I assume the battery can not be discharged to 0% or charged to 100%

Max is 100%. Minimum is 20% for a Powerwall (well, that's what the model uses—you can set your blackout limit to a lower number if you choose)

> How much should I take out when charging and discharging - is it a 10% loss each way?

I don't think we model that. If you really wanted to, you can use https://sam.nrel.gov/software-development-kit-sdk/pysam.html - though not sure that they have a Powerwall model. 

We're using that package in an internal simulation tool at the moment—is it complete overkill for our needs? Probably, but it meant I didn't have to write the logic for charging a battery.

> How to calculate a full cycle - if the battery is 50% charged, we add 50% to it and discharge 50% - should it be counted as 1/2 of a full cycle?

It's throughput, so yes, so that would be 1/2 a full cycle.

Smartshift is a linear optimisation model, so you could use something like EMHASS to come up with your own version: https://emhass.readthedocs.io/en/latest/

A bunch of people have some success with using that instead of Smartshift here.

**Reply by kryltsov** [reply]
Thanks!

**Reply by kryltsov** [reply]
I was thinking ...
API does not provide 5 min data I think only 30 min. So downloading my 2 years of data should not stress API. One can store those data locally (SQLLite etc) to model further and not abuse API. Perhaps you also use rate limiting anyway.

As you can see my report that was provided by Amber last year - uses 2 months of data in winter - so I doubt it is representative. I asked Amber again but I doubt they will take a bigger period. Besides, your support said all are busy onboarding standard utility companies 'refugees' right now :)

Maybe you guys can at least describe SmartShift logic so I can take it and build a Python script that loads historic data only once via API, and asks to enter the first bill data and last bill date (say my first bill is 15 Jan 2022 and last is 17 Aug 2023), the total amount paid in those bills, quoted Tesla installation and analyses data from/to date using SmartShift like logic on extended periods like 1-2 years to help Amber customers to decide if Tesla battery (which I understand the only one you support on SmartShift now) makes sense. Outputs will be - how much you would pay with SmartShift, and how many full cycles the battery completed (in other words - amortization will be used to reduce savings from solar)

**Reply by madpilot** [reply]
We do have an internal project to do what you are describing planned (we have done a trial for one network already), but it's a tough problem to solve at scale.

> So downloading my 2 years of data should not stress API.

It'll probably still time out as we price things on the fly for history to keep storage costs down. I'd do it month by month if you really want to do what you are trying to do.

> Maybe you guys can at least describe SmartShift logic so I can take it and build a Python script that loads historic data only once via API

We really don't have the time to manage, explain or support this. It's not a bunch of if statements, it's a somewhat complicated linear optimisation engine, that takes a bunch of inputs, has a bunch of constraints, which we are constantly tweaking. We just don't have the resources to document it to a way that an outsider would be able to grok it without a bunch of hand holding. And there isn't really a good business case for doing as such at this point. Sorry.

> Tesla battery (which I understand the only one you support on SmartShift now) makes sense

Worth nothing: We support 4 battery brands now: Tesla Powerwall, SolarEdge, Sungrow, and AlphaESS.

**Reply by kryltsov** [reply]
Thanks for the responses. In this case, I will just keep using solar panels, won't invest in battery and will give my other side project a priority :).

Just FYI I would pay $100 to be able to analyze my 2 years of data with batteries you support to see if a 14K investment into Tesla makes sense. Probably others will too. I am not a big spender but this represents a value and it can show me if I can save or earn more than $100 and do good for the $env :) 

If you can have 1K customers using this service a year - you will have 100K extra revenue to pay for computing and you can have another bullet point with a practical energy revolution statement.

Use data science interns they will queue to get experience with improving this for you for free :)

**Reply by iMortlake** [reply]
The future is uncertain. Humans are strange in that we generally use modelling to confirm what we've already decided. It's called 'confirmation bias'. I got the bug and just couldn't but it down until I justified getting my batteries. When I got my panels the pay-back was 3.5 years and then they changed the rules (FIT went way down). The batteries will take more like 6 years to pay back but I suspect the rules will change again before that happens. Electricity prices will change (up or down), FIT and subsidies will change, and cost of batteries will also change. No guarantees no mater how precise the modelling is. Good luck !

**Reply by kryltsov** [reply]
I feel the same - "cheated". You get "help" from the government to buy solar panels and then your export tariff goes negative. You export it and pay for it. Assuming the battery is like 14K vs. 11K for the solar panels - not going with them for now.

**Reply by chiefcomm** [reply]
Oh yes "cheated" is a good word :-) The export market in NSW spends so much time negative that Solar panels are beneficial only if you have a battery or an EV to suck up the excess and/or shift as much electric use during daylight hours and move usage outside the peaks.

**Comment by iMortlake** [comment]
Yes - but there are other options with no battery, normal retailers offer a fixed (low) FIT and VPPs offer a more reasonable fixed FIT.I agree it really schedule when the FIT regime changed so significantly after we bought solar. So I've since bought batteries and likely to get sting yet again down the track. The big question is what happens to prices and FIT even the gig coal generator closes. Lots of projects on the go but they will take time to get working. In the mean time I'm punting prices will rise further - but as always there's no guarantees.Sent from my Galaxy
-------- Original message --------From: chiefcomm ***@***.***> Date: 19/9/23  11:59 am  (GMT+10:00) To: amberelectric/public-api ***@***.***> Cc: iMortlake ***@***.***>, Comment ***@***.***> Subject: Re: [amberelectric/public-api] Battery assessment tool (Discussion
  #155) 
Oh yes "cheated" is a good word :-) The export market in NSW spends so much time negative that Solar panels are beneficial only if you have a battery or an EV to suck up the excess and/or shift as much electric use during daylight hours and move usage outside the peaks.
